Assessment/Assessment/ThreadSafeQueue_Impl.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include "ThreadSafeQueue.h"
+#include <mutex> // for unique_lock
+
+namespace Parallel
+{
+typedef std::unique_lock<std::shared_mutex> UniqueLock;
+typedef std::shared_lock<std::shared_mutex> SharedLock;
+
+
+template<typename T>
+ThreadSafeQueue<T>::ThreadSafeQueue(size_type maxSize)
+    : dataQueue_()
+    , mutex_()
+    , cVForProducers_()
+    , cVForConsumers_()
+    , reachedToMax_(false)
+    , maxSize_(maxSize)
+{}
+
+template<typename T>
+void
+ThreadSafeQueue<T>::push(const value_type v)
+{
+    UniqueLock lck(mutex_);
+
+    cVForProducers_.wait(lck, [this] {
+        if (dataQueue_.size() >= maxSize_)
+            reachedToMax_ = true;
+
+        return !reachedToMax_;
+    });
+
+    dataQueue_.push(v);
+
+    cVForConsumers_.notify_all();
+}
+
+
+template<typename T>
+void
+ThreadSafeQueue<T>::pop(value_type& v)
+{
+    UniqueLock lck(mutex_);
+
+    cVForConsumers_.wait(lck, [this] {
+        return !dataQueue_.empty();
+    });
+
+    v = dataQueue_.front();
+    dataQueue_.pop();
+
+    if (!reachedToMax_ && dataQueue_.size() < maxSize_)
```

Comment:
Is this condition really required? just notify producers and that is all.
dataQueue_.size() < maxSize_ is true in any case, because you are not allowing to insert above the limits.